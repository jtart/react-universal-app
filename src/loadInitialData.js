import { matchRoutes } from 'react-router-config';

const loadInitialData = async (url, routes) => {
  const matchedRoutes = matchRoutes(routes, url);

  if (matchedRoutes.length <= 0) {
    throw new Error(`No route was found for ${url}.`);
  }

  const { route, match } = matchedRoutes[0];

  if (!route.getInitialData) {
    return {};
  }

  try {
    return await route.getInitialData({ match });
  } catch (error) {
    throw error;
  }
};

export default loadInitialData;
