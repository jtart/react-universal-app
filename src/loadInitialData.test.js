import loadInitialData from './loadInitialData';
import * as reactRouterConfig from 'react-router-config';

describe('loadInitialData', () => {
  const data = 'Some data!';
  const match = { match: true };

  describe('no matched route', () => {
    it('should return an empty object', async () => {
      reactRouterConfig.matchRoutes = jest.fn().mockReturnValue([]);

      try {
        await loadInitialData('url', []);
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  });

  describe('no getInitialData function on route', () => {
    it('should return null', async () => {
      reactRouterConfig.matchRoutes = jest
        .fn()
        .mockReturnValue([
          { route: { route: 'data' }, match: { match: true } },
        ]);

      const initialData = await loadInitialData('url', ['url']);

      expect(initialData).toEqual(null);
    });
  });

  describe('getInitialData function on route', () => {
    it('should call getInitialData with passed ctx and return value', async () => {
      const route = {
        getInitialData: jest
          .fn()
          .mockImplementation(() => Promise.resolve(data)),
      };

      reactRouterConfig.matchRoutes = jest
        .fn()
        .mockReturnValue([{ route: route, match: match }]);

      const initialData = await loadInitialData('url', ['url']);

      expect(route.getInitialData).toHaveBeenCalledTimes(1);
      expect(route.getInitialData).toHaveBeenCalledWith({ match });
      expect(initialData).toEqual(data);
    });

    describe('error on getInitialData call', () => {
      it('should throw an error', async () => {
        const error = 'Error!';
        const route = {
          getInitialData: jest
            .fn()
            .mockImplementation(() => Promise.reject(error)),
        };

        reactRouterConfig.matchRoutes = jest
          .fn()
          .mockReturnValue([{ route: route, match: match }]);

        try {
          await loadInitialData(route, {});
        } catch (error) {
          expect(error).toEqual(error);
        }

        expect(route.getInitialData).toHaveBeenCalledTimes(1);
      });
    });
  });
});
