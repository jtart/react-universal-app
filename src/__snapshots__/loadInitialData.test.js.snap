// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadInitialData no matched route should return an empty object 1`] = `[Error: No route was found for url.]`;
